python/cuvs/cuvs/neighbors/hnsw/hnsw.pyx

```diff
@@ -0,0 +1,375 @@
+    """
+    cdef Index idx = Index()
+    cdef cuvsResources_t res = <cuvsResources_t>resources.get_c_obj()
+    cdef string c_filename = filename.encode('utf-8')
+    cdef cydlpack.DLDataType dl_dtype
+    if dtype == np.float32:
+        dl_dtype.code = cydlpack.kDLFloat
+        dl_dtype.bits = 32
+        dl_dtype.lanes = 1
+    elif dtype == np.ubyte:
+        dl_dtype.code = cydlpack.kDLUInt
+        dl_dtype.bits = 8
+        dl_dtype.lanes = 1
+    elif dtype == np.byte:
+        dl_dtype.code = cydlpack.kDLInt
+        dl_dtype.bits = 8
+        dl_dtype.lanes = 1
+    else:
+        raise ValueError("Only float32 is supported for dtype")
+
+    idx.index.dtype = dl_dtype
+    cdef cuvsDistanceType distance_type = DISTANCE_TYPES[metric]
+
+    check_cuvs(cuvsHnswDeserialize(
+        res,
+        c_filename.c_str(),
+        dim,
+        distance_type,
+        idx.index
+    ))
+    idx.trained = True
+    return idx
+
+
+@auto_sync_resources
+def from_cagra(cagra.Index index, resources=None):
+    """
+    Returns an hnsw base-layer-only index from a CAGRA index.
+
+    NOTE: This method uses the filesystem to write the CAGRA index in
+          `/tmp/<random_number>.bin` before reading it as an hnsw index,
+          then deleting the temporary file. The returned index is immutable
+          and can only be searched by the hnsw wrapper in cuVS, as the
+          format is not compatible with the original hnswlib library.
+          By `base_layer_only`, we mean that the hnsw index is created
+          without the additional layers that are used for the hierarchical
+          search in hnswlib. Instead, the base layer is used for the search.
+
+    Saving / loading the index is experimental. The serialization format is
+    subject to change.
+
+    Parameters
+    ----------
+    index : Index
+        Trained CAGRA index.
+    {resources_docstring}
+
+    Examples
+    --------
+    >>> import cupy as cp
+    >>> from cuvs.neighbors import cagra
+    >>> from cuvs.neighbors import hnsw
+    >>> n_samples = 50000
+    >>> n_features = 50
+    >>> dataset = cp.random.random_sample((n_samples, n_features),
+    ...                                   dtype=cp.float32)
+    >>> # Build index
+    >>> index = cagra.build(cagra.IndexParams(), dataset)
+    >>> # Serialize the CAGRA index to hnswlib base layer only index format
+    >>> hnsw_index = hnsw.from_cagra(index)
+    """
+    uuid_num = uuid.uuid4()
+    filename = f"/tmp/{uuid_num}.bin"
+    save(filename, index, resources=resources)
+    hnsw_index = load(filename, index.dim, np.dtype(index.active_index_type),
+                      "sqeuclidean", resources=resources)
+    os.remove(filename)
+    return hnsw_index
+
+
+@auto_sync_resources
+@auto_convert_output
+def search(SearchParams search_params,
+           Index index,
+           queries,
+           k,
+           neighbors=None,
+           distances=None,
+           resources=None):
+    """
```

Comment:
Will need to add HNSW to the docs for C, C++ and Python versions. We want to make sure this info is available for users (and that we can point to it in our getting started materials).